src/backends/backend/python/python_interpreter.cc

```diff
@@ -0,0 +1,902 @@
+      "can't get module path json value");
+
+  LOG_IF_ERROR(
+      module_path_json.MemberAsString("string_value", &module_path),
+      "can't get module path");
+
+  pymodule_path_ = module_path;
+  interpreter_pid_ = fork();
+
+  if (interpreter_pid_ == 0) {
+    // Use the python available in $PATH
+    // TODO: Make this overridable by config
+    std::string python_interpreter_path = "/usr/bin/python3";
+    std::string python_interpreter_startup =
+        "/workspace/builddir/server/install/lib/python/runtime/startup.py";
+
+    subinterpreter_commandline[0] = python_interpreter_path.c_str();
+    subinterpreter_commandline[1] = python_interpreter_startup.c_str();
+    subinterpreter_commandline[5] = pymodule_path_.c_str();
+    subinterpreter_commandline[7] = name_.c_str();
+    if (execve(
+            subinterpreter_commandline[0], (char**)subinterpreter_commandline,
+            nullptr) == -1) {
+      LOG_ERROR << "Cannot run interpreter host. Errno = " << errno << '\n'
+                << "python_interpreter_path: " << python_interpreter_path
+                << '\n'
+                << "python_interpreter_startup: " << python_interpreter_startup
+                << '\n'
+                << "pymodule_path_: " << pymodule_path_ << '\n'
+                << "instance_name: " << name_ << '\n';
+
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INVALID_ARG,
+          (std::string("Failed to initialize model instance ") + name_)
+              .c_str());
+    }
+  } else {
+    ConnectPythonInterpreter(module_path);
+  }
+
+  return nullptr;
+}
+
+int
+ModelInstanceState::ConnectPythonInterpreter(const std::string& module_path)
+{
+  auto grpc_channel =
+      grpc::CreateChannel(domain_socket_, grpc::InsecureChannelCredentials());
+  stub.reset(new ni::PythonInterpreter::Stub(grpc_channel));
+
+
+  std::shared_ptr<ni::InitializationCommand> initialization_params(
+      new ni::InitializationCommand());
+
+  std::vector<std::string> keys;
+  LOG_IF_ERROR(model_config.Members(&keys), "can't get key names");
+  std::string val;
+
+  const auto insert_model_param =
+      [&initialization_params](const std::string& val, const std::string& key) {
+        auto* value_pair = initialization_params->add_model_command();
+        value_pair->set_key(key);
+        value_pair->set_value(val);
+      };
+
+  insert_model_param("module_path", module_path);
+
+  // Attempting to connect to the python runtime
+  constexpr uint8_t conn_attempts = 5;
+  for (int i = 0; i < conn_attempts; ++i) {
+    grpc::ClientContext context;
+    ni::StatusCode result;
+    const auto err =
+        stub->InterpreterInit(&context, *initialization_params, &result);
+    if (err.ok() && result.code() == 0) {
+      return nic::ErrorCodes::Success;
+    } else {
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
+  }
+
+  std::this_thread::sleep_for(std::chrono::seconds(5));
```

Comment:
Why? And why is there an InterpreterInit in the loop and also another InterpreterInit below.

Reply:
Yes, they were not necessary. I removed them.